### PR TITLE
[FEATURE] Allow conditional expectations

### LINKED ISSUE #1657: Failed to save multiple Column Pair expectations
**Describe the bug**
Not possible to save more than 1 Column Pair expectation of the same expectation type. 

**To Reproduce**
1. Create an expectation suit for data source that supports Column Pair expectations (f.e. Pandas)
2. Add 2 Column Pair expectations with the same expectation type, but different columns values, f.e.
`batch.expect_column_pair_values_to_be_equal(column_A="Provider First Name", column_B="Provider Name Prefix Text")`
`batch.expect_column_pair_values_to_be_equal(column_A="Provider Last Name (Legal Name)", column_B="Provider Name Prefix Text")`
and run them.
3. Save the expectation suit
`batch.save_expectation_suite(discard_failed_expectations=False)`
4. Check generated json file with expectations and see 1 Column Pair expectation instead of 2.

**Expected behavior**
2 Column Pair expectations saved. 

**Environment**
 - OS: macOS Mojave 10.14.6
 - GE Version: 0.11.6

**Additional context**
As far as I get from code [here](https://github.com/great-expectations/great_expectations/blob/0.11.6/great_expectations/data_asset/data_asset.py#L411), it's possible to have multiple expectations of the same type only if `column`  field is provided. In the case of Column Pair expectations, there is no `column` field, but there are `column_A` and `column_B`. 
Also, this will be an issue for any custom expectation without the `column` field.

### PR DESCRIPTION
**Pull request by ArseniiDunaevQC**
This PR adds support for conditional expectations, which are used as an extension of the standard expectations. It does the following: adds a `condition` argument to all expectations, which one can use to narrow the scope of whatever expectations one is stating. For example, it is possible to expect values in some column to not be null _in general_ but only if other column or columns take on certain values.
It is only implemented for pandas dataset yet, and the `condition` argument should take a query string (expression) as stated in [pandas docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html).
The code also makes it possible to reflect conditional expectations in the Data Docs (e.g. in the following form: "if _condition_, then values must not be null").

### CODE REVIEW

**Comment by ArseniiDunaevQC** [comment]
There is still one thing I am unsure about, and I hope you can help me with it!
I have some doubts about the (in)equality tests for `ExpectationValidationResult`, since I am not sure what should `all()` in line 663 of _`great_expectations/core/__init__.py`_ do.
Some of my tests (namely, tests from _`tests/dataset/test_pandas_dataset_conditionals.py`_) were failing, when I tried to compare two `ExpectationValidationResult`s (with `!=` operator). In order for tests to run I implemented  the `__ne__` method for the class from above (lines 671-690), which is basically negation of the `__eq__` method, but without using the `all()` notation.

I am pretty sure that the tests are correct and that the `__ne__` method is redundant here, but I could not make tests pass otherwise. Could you please take a look and help me understand the logic behind the comparison of two `ExpectationValidationResult`s better?

Thank you so much in advance and of course I hope you will like my PR! :)

**Comment by kyleaton** [comment]
cla-bot check

**Comment by kyleaton** [comment]
@cla-bot check

File: great_expectations/dataset/dataset.py [hunk 1]
```diff
@@ -1252,8 +1255,6 @@ def expect_column_values_to_be_between(self,
                 If True, then catch exceptions and include them as part of the result object. \
                 For more detail, see :ref:`catch_exceptions`.
             meta (dict or None): \
-                A JSON-serializable dictionary (nesting allowed) that will be included in the output without \
```
Comment (krlng) [inline great_expectations/dataset/dataset.py]: Probably you removed this accidentaly

**Comment by krlng** [comment]
This is basically what I just did on my local branch and what I was thinking about to provide as a MR, but this one is actually way more profound. Sadly I don't understand the logic behind the testing-problem myself, but I would really love to see this one getting merged. Nice work @ArseniiDunaevQC

File: great_expectations/dataset/dataset.py (hunk 1 above)
Comment (ArseniiDunaevQC) [reply great_expectations/dataset/dataset.py]: Yep, you are completely right, thank you for reviewing!

**Comment by ArseniiDunaevQC** [comment]
Thank you @krlng! I am glad that there are users who already find this feature helpful!

File: docs/features/conditional_expectations.rst [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+########################
+Conditional Expectations
+########################
+
+Sometimes one may hold an Expectation not for a dataset in its entirety but only for a particular subset. Alternatively, what one expects of some variable may depend on the value of another.
+One may, for example, expect a column that holds the country of origin to not be null only for people of foreign descent.
+
+Great Expectations allows you to express such Conditional Expectations via a :code:`condition` argument that can be passed to all Dataset Expectations.
+
+Conditional Expectations are available only for the Pandas but not for the Spark and SQLAlchemy backends.
```
Comment (jcampbell) [inline docs/features/conditional_expectations.rst]:
```suggestion
Today, conditional Expectations are available only for the Pandas but not for the Spark and SQLAlchemy backends. The feature is **experimental**. Please expect changes to API as additional backends are supported.
```

File: docs/changelog/changelog.rst [hunk 3]
```diff
@@ -3,6 +3,7 @@
 develop
 -----------------
 * Remove the "project new" option from the command line (since it is not implemented; users can only run "init" to create a new project).
+* Add support for conditional expectations (works only for pandas dataset so far `#1217 <https://github.com/great-expectations/great_expectations/pull/1217>`_)
```
Comment (jcampbell) [inline docs/changelog/changelog.rst]:
```suggestion
* Add experimental support for conditional expectations (works only for pandas dataset so far `#1217 <https://github.com/great-expectations/great_expectations/pull/1217>`_)
```

**Review by jcampbell — COMMENTED** [review]
@ArseniiDunaevQC I am so excited about this huge contribution. Thank you!

I asked only for two small tweaks at this point to note that this API is experimental. We're moving toward making that a more formal designation. Let's signal it this way for now.

For the record, the thing that gaves me pause here is, as we discussed in email, the challenge of using pandas query language in the expectation configuration; this is the first really straining part of the expectation definition.

This does not need to block the PR, but some thoughts there:
 
1. We could potentially update the dataset_type key in the expectation suite to be "PandasDataset" instead of Dataset if the user sets a conditional in any expectation. That would signal that this suite is no longer compatible with SqlAlchemy and Spark.

2. On the other hand, it seems like it would be useful provide a SQL -> pandas query syntax converter. Given the parser that you have included for the renderer (wow!) it seems like it would be feasible to do something fairly straightforward there, starting with essentially just translating `=` => `==`. Since I have now included an expression language parser in the evaluation expressions this also seems more in scope now than it has before. What do you think of that as an option?

Again, nothing needs to block this PR from my perspective. I want to flag for my colleague @eugmandel to take a look from the perspective of making sure that we are getting the 'experimental' signaling correct as well, and pending his take and your views on the above I think we should move on forward!

**Comment by ArseniiDunaevQC** [comment]
Thank you @jcampbell for the review!

Sure, let's mark this feature __experimental__ for now.

Regarding the language of conditional expectations: just in case I understand you right, you would like condition syntax to be more SQL-like (or even 'native' SQL)? In this case the second option seems feasible to me, since I think it would be useful in future to provide support for the remaining datasets (sparkdf and sqlalchemy).

Before the further discussion on the feature I just wanted to make sure you have seen my question about the testing issues from the above https://github.com/great-expectations/great_expectations/pull/1217#issuecomment-603467018 as I am still a little bit concerned about this.

**Comment by jcampbell** [comment]
I reviewed some related approaches to addressing this, and would like to propose the following:

1. we *not* gate on any sort of change to the parsing, but:
2. we build in future flexibility by:
  a. adding a new "condition_engine" kwarg with the default parameter "pandas"
  b. rename "condition" to "row_condition"

I *think* this a really minor line-in line-out change that will make this rock and go as is!

**Comment by ArseniiDunaevQC** [comment]
Great idea @jcampbell! It makes sense for me and I think it will work for now.

So, I made the changes you proposed, but got stuck a little bit on appending new expectations. Given the #1297 and `appliesToSameData`, I suppose it is better dedicated on your side (in future). As for now, I decided that having different `condition_engine`s means applying to non-identical data, and thus, different expectations.
Anyway, it is not really relevant for the moment since we only have one default `condition_engine` (namely, "pandas"), but it could be hard in future to distinguish between two conditional expectations with (potentially) same `row_condition`s, but distinct engines.

**Comment by jcampbell** [comment]
Perfect -- I think your assessment is right.

Two really small things then: 1. have you checked on the failing tests with old pandas? 2. Would you sign the CLA?

**Comment by jcampbell** [comment]
Love this PR conceptually, but absent CLA we're going to close for now.

**Comment by krlng** [comment]
Oh no, really? @ArseniiDunaevQC

**Comment by xhochy** [comment]
@jcampbell Can you reopen? The CLA should be signed by now.

**Comment by jcampbell** [comment]
@cla-bot check

**Review by jcampbell — APPROVED** [review]
SO EXCITED FOR THIS FEATURE!!!!

**Pull request merged**